src/vanilla/server.rs

```diff
@@ -0,0 +1,337 @@
+
+use rand;
+use time;
+use uuid::Uuid;
+
+const PACKET_NAMES: [&'static str; 26] = [
+    "(c2s) KeepAlive",
+    "(c2s) ChatMessage",
+    "(c2s) UseEntity",
+    "(c2s) Player",
+    "(c2s) PlayerPosition",
+    "(c2s) PlayerLook",
+    "(c2s) PlayerPositionAndLook",
+    "(c2s) PlayerDigging",
+    "(c2s) PlayerBlockPlacement",
+    "(c2s) HeldItemChange",
+    "(c2s) Animation",
+    "(c2s) EntityAction",
+    "(c2s) SteerVehicle",
+    "(c2s) CloseWindow",
+    "(c2s) ClickWindow",
+    "(c2s) ConfirmTransaction",
+    "(c2s) CreativeInventoryAction",
+    "(c2s) EnchantItem",
+    "(c2s) UpdateSign",
+    "(c2s) PlayerAbilities",
+    "(c2s) TabComplete",
+    "(c2s) ClientSettings",
+    "(c2s) ClientStatus",
+    "(c2s) PluginMessage",
+    "(c2s) Spectate",
+    "(c2s) ResourcePackStatus"
+];
+
+/// TODO(toqueteos): Move this to its own module. Proposal: src/vanilla/mod.rs
+pub struct Server {
+    addr: String,
+    props: Properties,
+    // Dummy player storage, just their username.
+    players: Vec<String>,
+    // World start timestamp
+    start: time::Timespec
+}
+
+impl Server {
+    pub fn new() -> io::Result<Server> {
+        let properties_path = &Path::new("server.properties");
+        let props = if properties_path.exists() {
+            try!(Properties::load(properties_path))
+        } else {
+            Properties::default()
+        };
+        info!("{:?}", props);
+
+        // There's no *prettier way* of doing this, if it was an Option then
+        // there's .unwrap_or but it's just a String.
+        let addr = if props.server_ip.is_empty() {
+            "0.0.0.0".to_string()
+        } else {
+            props.server_ip.clone()
+        };
+        Ok(Server {
+            addr: addr,
+            props: props,
+            players: Vec::new(),
+            start: time::get_time()
+        })
+    }
+
+    pub fn addr(&self) -> &str { return &self.addr }
+    pub fn port(&self) -> u16 { self.props.server_port }
+
+    // FIXME(toqueteos): Read from world's level.dat file
+    pub fn world_age(&self) -> i64 {
+        let end = time::get_time();
+        let elapsed = end.sub(self.start).num_seconds();
+        elapsed * 20
+    }
+
+    // FIXME(toqueteos): Read from world's level.dat file
+    pub fn time_of_day(&self) -> i64 {
+        self.world_age() % 24000
+    }
+
+    pub fn handle(&self, mut stream: TcpStream) -> io::Result<()> {
+        let state = try!(self.handle_handshake(&mut stream));
+        let client = try!(stream.peer_addr());
+        match state {
+            NextState::Status => {
+                try!(slp::response(&mut stream));
+                try!(slp::pong(&mut stream));
+            }
+            NextState::Login => {
+                try!(self.handle_login(&mut stream));
+                try!(self.handle_post_login(&mut stream));
+                // <Insert game loop here>
+                // self.game_loop(&mut stream);
+            }
+        };
+        Ok(())
+    }
```

Comment:
`handle_handshake`, `handle_login`, and `handle_post_login` still need to be moved into this function. Also it would be better for the `NextState::Status` branch to loop over incoming packets and handle pings and status requests in any order.